python/ray/util/ray_executor.py

```diff
@@ -0,0 +1,317 @@
+    List,
+    Any,
+    TypeVar,
+    ParamSpec,
+    Iterable,
+    Iterator,
+    TYPE_CHECKING,
+)
+
+import ray
+from ray.util.actor_pool import ActorPool
+from ray.util.annotations import PublicAPI
+
+
+# Typing -----------------------------------------------
+
+if TYPE_CHECKING:
+    from ray._private.worker import BaseContext, RemoteFunction0
+
+T = TypeVar("T")
+P = ParamSpec("P")
+
+# ------------------------------------------------------
+
+
+@PublicAPI(stability="alpha")
+class RayExecutor(Executor):
+    """`RayExecutor` is a drop-in replacement for `ProcessPoolExecutor` and
+    `ThreadPoolExecutor` from `concurrent.futures` but distributes and executes
+    the specified tasks over a Ray cluster instead of multiple processes or
+    threads.
+
+    Args:
+        max_workers:
+            If max_workers=None, the work is distributed over the number of
+            CPUs available in the cluster (num_cpus) , otherwise the number of
+            CPUs is limited to the value of max_workers.
+
+        All additional keyword arguments are passed to ray.init()
+        (see https://docs.ray.io/en/latest/ray-core/package-ref.html#ray-init).
+
+        For example, this will connect to a cluster at the specified address:
+        .. code-block:: python
+
+            RayExecutor(address='192.168.0.123:25001')
+
+        Note: excluding an address will initialise a local Ray cluster.
+    """
+
+    def __init__(
+        self, max_workers: Optional[int] = None, shutdown_ray: bool = True, **kwargs
+    ):
+
+        """
+        Initialise a new RayExecutor instance which distributes tasks over
+        a Ray cluster.
+
+        self._shutdown_ray:
+            If False, the Ray cluster is not destroyed when self.shutdown() is
+            called. This prevents initialising a new cluster every time the
+            executor is used in a `with` context. Futures will be
+            destroyed regardless of the value of `_shutdown_ray`.
+        self._shutdown_lock:
+            This is set to True once self.shutdown() is called. Further task
+            submissions are blocked.
+        self._futures:
+            Futures are aggregated into this list as they are returned.
+        self.__actor_pool:
+            ActorPool which is used by self.map() when max_workers > 1
+        self.__remote_fn:
+            Wrapper around the remote function to be executed by Ray.
+        self.context:
+            Context containing settings and attributes returned after
+            initialising the Ray client.
+        """
+
+        self._shutdown_ray: bool = shutdown_ray
+        self._shutdown_lock: bool = False
+        self._futures: List[Future] = []
+        self.__actor_pool: Optional[ActorPool] = None
+        self.__remote_fn: "Optional[RemoteFunction0]" = None
+        self.context: "Optional[BaseContext]" = None
+
+        # The following is necessary because `@ray.remote` is only available at runtime.
+        if max_workers is None:
```

Comment:
we should always use actor based approach. When max_workers is None, the # actors should equal to the number of cpus in the cluster.

`multiprocessing.Pool` does this as well: https://github.com/ray-project/ray/blob/master/python/ray/util/multiprocessing/pool.py#L633

Reply:
I've chosen the opposite approach: we always simply use `num_cpus` and have removed actors.